cedar-policy-validator/src/custom_schema/err.rs

```diff
@@ -0,0 +1,322 @@
+use std::{
+    collections::HashMap,
+    fmt::Display,
+    ops::{Deref, DerefMut},
+};
+
+use cedar_policy_core::parser::{err::expected_to_string, unescape::UnescapeError, Loc, Node};
+use lalrpop_util as lalr;
+use lazy_static::lazy_static;
+use miette::{Diagnostic, LabeledSpan, SourceSpan};
+use nonempty::NonEmpty;
+use smol_str::SmolStr;
+use thiserror::Error;
+
+use super::ast::Str;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum UserError {
+    EmptyList,
+    StringEscape(NonEmpty<UnescapeError>),
+}
+
+pub(crate) type RawLocation = usize;
+pub(crate) type RawToken<'a> = lalr::lexer::Token<'a>;
+pub(crate) type RawUserError = Node<UserError>;
+
+pub(crate) type RawParseError<'a> = lalr::ParseError<RawLocation, RawToken<'a>, RawUserError>;
+pub(crate) type RawErrorRecovery<'a> = lalr::ErrorRecovery<RawLocation, RawToken<'a>, RawUserError>;
+
+type OwnedRawParseError = lalr::ParseError<RawLocation, String, RawUserError>;
+
+lazy_static! {
+    /// Keys mirror the token names defined in the `match` block of
+    /// `grammar.lalrpop`.
+    static ref FRIENDLY_TOKEN_NAMES: HashMap<&'static str, &'static str> = HashMap::from([
+        ("IN", "`in`"),
+        ("PRINCIPAL", "`principal`"),
+        ("ACTION", "`action`"),
+        ("RESOURCE", "`resource`"),
+        ("CONTEXT", "`context`"),
+        ("STRINGLIT", "string literal"),
+        ("ENTITY", "`entity`"),
+        ("NAMESPACE", "`namespace`"),
+        ("TYPE", "`type`"),
+    ]);
+}
+
+/// For errors during parsing
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ParseError {
+    /// Error generated by lalrpop
+    ToAST(OwnedRawParseError),
+}
+
+impl ParseError {
+    pub(crate) fn from_raw_parse_err(err: RawParseError<'_>) -> Self {
+        Self::ToAST(err.map_token(|token| token.to_string()))
+    }
+
+    pub(crate) fn from_raw_err_recovery(recovery: RawErrorRecovery<'_>) -> Self {
+        Self::from_raw_parse_err(recovery.error)
+    }
+
+    /// Extract a primary source span locating the error.
+    pub fn primary_source_span(&self) -> SourceSpan {
+        let Self::ToAST(err) = self;
+        match err {
+            OwnedRawParseError::InvalidToken { location } => SourceSpan::from(*location),
+            OwnedRawParseError::UnrecognizedEof { location, .. } => SourceSpan::from(*location),
+            OwnedRawParseError::UnrecognizedToken {
+                token: (token_start, _, token_end),
+                ..
+            } => SourceSpan::from(*token_start..*token_end),
+            OwnedRawParseError::ExtraToken {
+                token: (token_start, _, token_end),
+            } => SourceSpan::from(*token_start..*token_end),
+            OwnedRawParseError::User { error } => error.loc.span,
+        }
+    }
+}
+
+impl Display for ParseError {
```

Comment:
can we define this using `thiserror`'s `#[error]` macros, like we do for other error types in our codebase